pkg/logs/tag/provider.go

```diff
@@ -28,32 +30,70 @@ var NoopProvider Provider = &noopProvider{}
 
 // provider caches a list of up-to-date tags for a given entity polling periodically the tagger.
 type provider struct {
-	entityID string
-	tags     []string
-	done     chan struct{}
-	mu       sync.Mutex
+	entityID             string
+	tags                 []string
+	done                 chan struct{}
+	forceRefresh         bool
+	forceRefreshDuration time.Duration
+	taggerWarmupDuration time.Duration
+	mu                   sync.Mutex
 }
 
 // NewProvider returns a new Provider.
 func NewProvider(entityID string) Provider {
-	return &provider{
-		entityID: entityID,
-		tags:     []string{},
-		done:     make(chan struct{}),
+	p := &provider{
+		entityID:     entityID,
+		tags:         []string{},
+		done:         make(chan struct{}),
+		forceRefresh: config.Datadog.GetBool("logs_config.k8s_wait_for_tags"),
 	}
+	if p.forceRefresh {
+		p.forceRefreshDuration = config.Datadog.GetDuration("logs_config.force_tagger_call_duration") * time.Second
+		p.taggerWarmupDuration = config.Datadog.GetDuration("logs_config.tagger_warmup_duration") * time.Second
+	}
+	return p
 }
 
 // GetTags returns the list of up-to-date tags.
 func (p *provider) GetTags() []string {
+	if p.getForceRefresh() {
+		p.updateTags()
+	}
 	p.mu.Lock()
 	defer p.mu.Unlock()
 	return p.tags
 }
 
+// setForceRefresh udpates the forceRefresh attribute
+func (p *provider) setForceRefresh(b bool) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.forceRefresh = b
+}
+
+// getForceRefresh returns the forceRefresh attribute
+func (p *provider) getForceRefresh() bool {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	return p.forceRefresh
+}
+
 // Start starts the polling of new tags on another go routine.
 func (p *provider) Start() {
+	// Warmup duration: make sure the tagger collects all the service tags
+	<-time.After(p.taggerWarmupDuration)
+	// Initial tags update
+	p.updateTags()
```

Comment:
these 2 lines will change the function from being entirely non-blocking to being blocking for 1 second + waiting until `updateTags` completes. Is that OK for the caller, or could this have undesirable effects upstream?

Reply:
the side effect is having the first few logs a bit late - but correctly tagged, our solutions engineers will communicate that to customers when suggesting this fix.

Reply:
Have you checked whether this affects other tailers or sources as well, on the caller? I'm not familiar with the logs codebase but in general I'd be cautious around adding delays in methods that the caller assumes are non-blocking.

In the same vein, to keep `Start` non-blocking unless `logs_config.k8s_wait_for_tags` is enabled, can we keep `updateTags` in the goroutine started below?

Reply:
yes it affects other tailers but I do believe it's important to **ensure** we populate the local cache at least one time before calling the `GetTags` method (by calling `updateTags` before the `Start` method returns

Reply:
hmm, to make sure I understand, are you concerned about the time it would take for the 1st call to `GetTags` to complete if the call it makes to `updateTags` has to populate the local cache?

Reply:
My concern is when `k8s_wait_for_tags` is disabled, `GetTags` wouldn't be forced to call `updateTags` and returns the local cache directly. In this case there no guarantee the local cache has been populated.

Reply:
ok I see, so to summarize:

* when `k8s_wait_for_tags` is disabled (default), we want to be absolutely sure the local cache is populated before `GetTags` is called, hence the blocking call to `updateTags` here.
* when `k8s_wait_for_tags` is enabled, we wait for 1 second (default) before this initial update, and then update on every call to `GetTags` for 5 seconds (default), to make sure we also get any new tags that the AD may find for this kubelet.

To be honest I don't what the side effects of making `Start` blocking are, but could you update the comment documenting the `Start` function, to reflect that it does a first blocking update of tags? And also update the PR description to reflect that this PR introduces this small change even with `k8s_wait_for_tags` left to its default value of `false` :)

Reply:
That's correct, thanks for summarizing 👍 
I documented the `Start` method and updating the PR description to reflect that, and I also moved the wait for tagger duration when `k8s_wait_for_tags` is enabled) to the `GetTags` method instead of `Start`. This way we only block forwarding messages (in `tailer.forwardMessages()` instead of blocking the whole Start logic in the scheduler.